File: djangobase/basiswurzel.py
# -*- coding: utf-8 -*-
"""Unter welchem Praefix `djangobase.urls` eingebunden ist.

WARUM ES DIESE KLASSE GIBT (Befund 27.08.2026, 3DTools)
=======================================================
Jedes Projekt bindet `djangobase.urls` unter einem eigenen Praefix ein::

    shortlongx   path("hilfe/", include("djangobase.urls"))   -> /hilfe/...
    assistant    path("",       include("djangobase.urls"))   -> /...
    3DTools      path("help/",  include("djangobase.urls"))   -> /help/...

Vier mitgelieferte JS-Module hatten `/hilfe/tests/aufzeichnung/` fest im Text.
In 3DTools liefen sie damit auf **404 bei jedem Seitenaufruf** — dreimal je
Seite, ohne Fehlerseite, ohne Eintrag im Fehlerlog. Gefunden wurde es erst
durch eine Browserprobe, die auf Antwortcodes >= 400 achtet.

`SystemStatsLeiste.URL` loeste dasselbe Problem, indem es die Adresse
UEBERSCHREIBBAR machte — die Last liegt damit beim Projekt, und wer den Schritt
vergisst, hat wieder eine stille 404. Hier steht die Wurzel stattdessen EINMAL
im Grundgeruest, und die Module lesen sie von dort.
"""

from django.urls import NoReverseMatch, reverse


class Basiswurzel:
    """Der Praefix, unter dem `djangobase.urls` haengt — mit Schrägstrich."""

    #: Route, ueber die rueckwaerts gerechnet wird. Sie ist seit der ersten
    #: Fassung dabei; faellt sie je weg, meldet der Test `basiswurzel` das.
    ANKER = "djangobase:versionen"

    #: Was `ANKER` hinter der Wurzel anhaengt.
    ANKERWEG = "versionen/"

    #: Was gilt, wenn `djangobase.urls` gar nicht eingebunden ist. Historischer
    #: Wert — die Mehrzahl der Projekte haengt dort.
    ERSATZ = "/hilfe/"

    @staticmethod
    def weg():
        """@returns z. B. ``'/help/'``, ``'/hilfe/'`` oder ``'/'``."""
        try:
            adresse = reverse(Basiswurzel.ANKER)
        except NoReverseMatch:
            return Basiswurzel.ERSATZ
        if not adresse.endswith(Basiswurzel.ANKERWEG):
            # Ein Projekt hat die Route umgehaengt. Dann ist der Rueckschluss
            # auf die Wurzel nicht mehr gueltig — lieber der alte Wert als ein
            # falsch abgeschnittener Pfad.
            return Basiswurzel.ERSATZ
        return adresse[:-len(Basiswurzel.ANKERWEG)]

File: djangobase/static/djangobase/js/basiswurzel.js
/**
 * Unter welchem Praefix `djangobase.urls` eingebunden ist.
 *
 * WARUM (Befund 27.08.2026, 3DTools)
 * ==================================
 * Jedes Projekt haengt djangoBase woanders ein:
 *
 *     shortlongx   /hilfe/...
 *     assistant    /...
 *     3DTools      /help/...
 *
 * `aufzeichner.js`, `aufzeichner_abspieler.js`, `aufzeichner_leiste.js` und
 * `aufzeichnung.js` hatten `/hilfe/tests/aufzeichnung/` fest im Text. In
 * 3DTools liefen sie damit bei JEDEM Seitenaufruf dreimal in eine 404 — keine
 * Fehlerseite, kein Eintrag im Fehlerlog, die Aufnahme war einfach still tot.
 *
 * Die Wurzel steht seither einmal im Grundgeruest (`_shell.html`,
 * `<meta name="djangobase-wurzel">`) und wird hier gelesen.
 */
export class Basiswurzel {
    /** Name des Kennzeichens im Kopf der Seite. */
    static KENNZEICHEN = 'djangobase-wurzel';

    /** Was gilt, wenn das Kennzeichen fehlt — die haeufigste Einbindung. */
    static ERSATZ = '/hilfe/';

    /** @returns {string} z. B. `'/help/'`, `'/hilfe/'` oder `'/'` */
    static wurzel() {
        const kopf = document.querySelector(
            `meta[name="${Basiswurzel.KENNZEICHEN}"]`);
        const wert = kopf?.getAttribute('content') || '';
        // Ein leeres `content` ist kein gueltiger Praefix: Es entstuende
        // `tests/aufzeichnung/` — relativ zur aktuellen Seite, also woanders
        // je nachdem, wo man gerade steht.
        if (!wert.startsWith('/')) return Basiswurzel.ERSATZ;
        return wert.endsWith('/') ? wert : wert + '/';
    }

    /**
     * Vollstaendige Adresse einer djangoBase-Route.
     * @param {string} rest Weg hinter der Wurzel, z. B. `'tests/aufzeichnung/'`
     * @returns {string}
     */
    static weg(rest) {
        return Basiswurzel.wurzel() + String(rest).replace(/^\/+/, '');
    }
}
